--- a/drivers/iommu/apple-dart.c
+++ b/drivers/iommu/apple-dart.c
@@ -364,9 +364,11 @@
 static void apple_dart_s5l8960x_write_tcr(struct apple_dart *dart, u32 sid, u32 val)
 {
-	WARN_ON(sid >= 4);
-	u32 tcr = readl(dart->regs + dart->hw->tcr);
+	u32 tcr;
 
-	tcr &= 0xff << (sid * DART_S5L8960X_TCR_BITS_PER_STREAM);
-	tcr |= val;
+	if (WARN_ON(sid >= 4 || val > 0xff))
+		return;
+	tcr = readl(dart->regs + dart->hw->tcr);
+	tcr &= ~(0xffU << (sid * DART_S5L8960X_TCR_BITS_PER_STREAM));
+	tcr |= val << (sid * DART_S5L8960X_TCR_BITS_PER_STREAM);
 	writel(tcr, dart->regs + dart->hw->tcr);
 }
